### PR TITLE
Adopt more FindMPI and requires CMake 3.10

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
~~#2286 is needed for this PR.~~
This PR use FindMPI even if the CMAKE_C/CXX_COMPILER are already compiler wrappers.
1. FindMPI is significantly improved in cmake 3.9 and 3.10 deprecated MPIEXEC and prefers MPIEXEC_EXECUTABLE. So this PR requires cmake 3.10 which is released on Nov 20th 2017.
2. Users configure and compile as usual. The logic how FindMPI finds MPI is provided in https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/FindMPI.cmake#L101
Most user should hit case 2 and we issue warning in case 3 as before.
3. It becomes possible to add library specific settings, I'm able to detect MVAPICH2 and issue warnings for #1703. We can also add OpenMPI specific flags to fix the slow testing problem with threads in the future.
4. FindMPI defines MPI::MPI_CXX target (modern cmake style) I use it in our dependency tree.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## What systems has this change been tested on?
ALCF Theta Intel
NERSC Cori Intel
OLCF Summit GNU, XL, CUDA
My workstation, GNU no MPI, GNU finds MPI not via compiler wrapper.

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [ ] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [x] documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
@prckent updated the manual and added backward compatibility with warning.

**Comment by ye-luo** [comment]
Test this please

File: CMakeLists.txt [hunk 1]
```diff
@@ -540,30 +495,52 @@ IF (BUILD_AFQMC AND NOT APPLE)
...
+  IF(${MPI_CXX_LIBRARY_VERSION_STRING} MATCHES "MVAPICH2" AND NOT ${MPI_CXX_LIBRARY_VERSION_STRING} MATCHES "disable-registration-cache")
+    MESSAGE(FATAL_ERROR "MVAPICH2 with registration cache enabled breaks QMCPACK. "
+                        "Use a different MPI library or build MVAPICH2 with --disable-registration-cache configure option.")
   ENDIF()
 
-  IF ( NOT DEFINED MPIEXEC_PREFLAGS )
-    SET( MPIEXEC_PREFLAGS "" )
+  IF(NOT ${CMAKE_CXX_COMPILER} STREQUAL ${MPI_CXX_COMPILER})
+    SET(MPI_WARNING_LIST
+        "Building MPI version without using MPI compiler wrappers.\n"
```
Comment (markdewing) [inline CMakeLists.txt]: I think the preferred way to use CMake with MPI is to set the the compiler variables to the underlying compiler, and let FindMPI find all the include paths and libraries, etc.   Is there something in particular that we know will break QMCPACK (like not all the include or libraries paths are specified in a way that will pick up the MPI files?).     I don't think anything should be done on this PR, just pointing it out here.

Comment (prckent) [reply CMakeLists.txt]: The safe way to use MPI with cmake is to use the MPI compiler wrappers. This give a guaranteed outcome and the correct libraries will be linked. We recommend this route in the manual. Mileage may vary the other way, particularly on a "legacy" machine with multiple dusty MPI installations.

**Comment by prckent** [comment]
I tested a subset of the nightlies using Intel MPI, OpenMPI, and no MPI. These all configured and ran indicating that the likely use cases for the FindMPI implementation are functional.

File: CMakeLists.txt (hunk 1 above)
Comment (ye-luo) [reply CMakeLists.txt]: IMO, the preferred way is passing MPI compiler wrappers as compilers. FindMPI knows how to deal with this and keep the compile/link line clean. If the compilers passed in are not compiler wrappers, FindMPI goes to the folder containing MPIEXEC_EXECUTABLE and look for compiler wrappers of all the needed languages. Although it probably can find all the libraries and set proper flags and include directories based on querying MPIEXEC_EXECUTABLE options like -echo/-showme, the resulting compile and link lines get extremely long and error prone.

**Pull request merged**